packages/flutter/lib/src/rendering/sliver.dart

```diff
@@ -1742,46 +1440,599 @@ class RenderViewport2 extends RenderBox with ContainerRenderObjectMixin<RenderSl
+    final SliverPhysicalParentData childParentData = child.parentData;
+    childParentData.applyPaintTransform(transform);
+  }
+
+  @override
+  double computeChildMainAxisPosition(RenderSliver child, double parentMainAxisPosition) {
+    assert(child != null);
+    assert(child.constraints != null);
+    final SliverPhysicalParentData childParentData = child.parentData;
+    switch (applyGrowthDirectionToAxisDirection(child.constraints.axisDirection, child.constraints.growthDirection)) {
+      case AxisDirection.up:
+        return child.geometry.paintExtent - (parentMainAxisPosition - childParentData.paintOffset.dy);
+      case AxisDirection.right:
+        return parentMainAxisPosition - childParentData.paintOffset.dx;
+      case AxisDirection.down:
+        return parentMainAxisPosition - childParentData.paintOffset.dy;
+      case AxisDirection.left:
+        return child.geometry.paintExtent - (parentMainAxisPosition - childParentData.paintOffset.dx);
+    }
+    return 0.0;
+  }
+
+  @override
+  void rereportDimensions() {
+    assert(_minScrollExtent != null);
+    assert(_maxScrollExtent != null);
+    assert(anchor != null);
+    double effectiveExtent;
+    switch (axis) {
+      case Axis.vertical:
+        effectiveExtent = size.height;
+        break;
+      case Axis.horizontal:
+        effectiveExtent = size.width;
+        break;
+    }
+    assert(effectiveExtent != null);
+    offset.applyViewportDimension(effectiveExtent);
+    if (!offset.applyContentDimensions(
+          // when updating this, also update similar code in performLayout()
+          math.min(0.0, _minScrollExtent + effectiveExtent * anchor),
+          math.max(0.0, _maxScrollExtent - effectiveExtent * (1.0 - anchor)),
+       ))
+     markNeedsLayout();
+  }
+
+  @override
+  int indexOfFirstChild() {
+    assert(center != null);
+    assert(center.parent == this);
+    assert(firstChild != null);
     int count = 0;
     RenderSliver child = center;
     while (child != firstChild) {
       count -= 1;
       child = childBefore(child);
     }
+    return count;
+  }
 
-    child = firstChild;
-    while (child != lastChild) {
-      result += '${child.toStringDeep("$prefix \u251C\u2500${_labelChild(count)}: ", "$prefix \u2502")}';
-      count += 1;
+  @override
+  String labelChild(int index) {
+    if (index == 0)
+      return 'center child';
+    return 'child $index';
+  }
+
+  @override
+  Iterable<RenderSliver> get childrenInPaintOrder sync* {
+    if (firstChild == null)
+      return;
+    RenderSliver child = firstChild;
+    while (child != center) {
+      yield child;
       child = childAfter(child);
     }
-    if (child != null) {
-      assert(child == lastChild);
-      result += '${child.toStringDeep("$prefix \u2514\u2500${_labelChild(count)}: ", "$prefix  ")}';
+    child = lastChild;
+    do {
+      yield child;
+      if (child == center)
+        return;
+      child = childBefore(child);
+    } while (true);
```

Comment:
I generally prefer `while (true) { ... }` infinite loops, but maybe that's not in the style guide (yet).